### PR TITLE
ARROW-7740: [C++] Fix StructArray::Flatten corruption

### PR DESCRIPTION
**Pull request by fsaintjacques**
Flatten would modify in-place the data_->child_data ArrayData if the currently flattened array is a slice. The parent's child_data values would also be modified (and corrupted with wrong offset/length).

### CODE REVIEW

File: r/R/json.R [hunk 1]
```diff
@@ -38,6 +38,8 @@
 read_json_arrow <- function(file, col_select = NULL, as_data_frame = TRUE, ...) {
   tab <- JsonTableReader$create(file, ...)$Read()$select(!!enquo(col_select))
 
+  tab$ValidateFull()
```
Comment (fsaintjacques) [inline r/R/json.R]: @nealrichardson should we toggle this via an option?

Comment (nealrichardson) [reply r/R/json.R]: I don't have an opinion; sounds like you think we should? Do any of the other readers do validation like this at all? (I'm not aware that they do.)

Comment (nealrichardson) [reply r/R/json.R]: Oh I see, this is added in this patch. IMO the library shouldn't be returning invalid Tables, so calling `Validate` on them seems unnecessary. I wouldn't add this to the reader function here; instead, if someone wanted to validate, they could return the Table themselves (`as_data_frame = FALSE`) and then call `$ValidateFull()` on it themselves.

File: r/tests/testthat/test-json.R [hunk 2]
```diff
@@ -147,3 +147,16 @@ test_that("Can read json file with nested columns (ARROW-5503)", {
...
+  tf <- tempfile()
+  on.exit(unlink(tf))
+  writeLines('
+    {"a":[{"b":1.0},{"b":2.0}]}
+    {"a":[{"b":1.0},{"b":2.0}]}
+  ', tf)
+
+  one <- tibble::tibble(b = c(1, 2))
+  expected <- tibble::tibble(a = c(list(one), list(one)))
+  expect_equivalent(arrow::read_json_arrow(tf), expected)
```
Comment (nealrichardson) [inline r/tests/testthat/test-json.R]:
```suggestion
  expect_equivalent(read_json_arrow(tf), expected)
```

**Review by nealrichardson — CHANGES_REQUESTED** [review]

File: r/R/json.R (hunk 1 above)
Comment (wesm) [reply r/R/json.R]: I would remove this, it was only to help me debug

**Comment by wesm** [comment]
Where was Flatten being called in the original failing test case?

**Comment by fsaintjacques** [comment]
Flatten is called by the Array -> R vector  struct [converter](https://github.com/apache/arrow/blob/master/r/src/array_to_vector.cpp#L395-L407).

**Review by wesm — APPROVED** [review]
+1 -- aside from the `tab$ValidateFull()` in json.R this is good. I'll remove that and then merge after CI runs again

**Pull request closed without merging**